### PR TITLE
[SPARK-38041][SQL] DataFilter pushed down dynamically

### PR DESCRIPTION
**Pull request by jackylee-ch**
### Why are the changes needed?

At present, the Filter is divided into DataFilter and PartitionFilter when it is pushed down, and use PartitionFilter to prune partition and use DataFilter to prune data. However, the pruned down DataFilter contains all data filter for all partitions, not for one specific partition, which may lead to a bad behavior such as scan all data of each partition.
This pr will push down dataFilter with PartitionFilter, and evaluate and prune partitionFilter at runtime before push to parquet, which can push less filter to parquet.

Here is a example.
```scala
val df = spark.range(0, 1000).withColumnRenamed("id", "a").withColumn("b", $"a")
df.withColumn("c", lit(0)).write.mode("overwrite").partitionBy("c").parquet("/tmp")
df.withColumn("c", lit(1)).write.mode("append").partitionBy("c").parquet("/tmp")
df.withColumn("c", lit(2)).write.mode("append").partitionBy("c").parquet("/tmp")
spark.read.parquet("/tmp").filter("(a < 10 AND c = 0) OR (a >= 10 AND c >= 1 AND c < 3)").explain(true)
```

before
DataFilters pushed to ParquetScan is `[Or(LessThan(a,10),GreaterThanOrEqual(a,10))]`
Filters pushed to ParquetInputFormat is `[Or(LessThan(a,10),GreaterThanOrEqual(a,10))]`
```
== Physical Plan ==
*(1) Filter (((a#40L < 10) AND (c#42 = 0)) OR (((a#40L >= 10) AND (c#42 >= 1)) AND (c#42 < 3)))
+- *(1) ColumnarToRow
   +- BatchScan[a#40L, b#41L, c#42, d#43] ParquetScan DataFilters: [(((a#40L < 10) AND (c#42 = 0)) OR (((a#40L >= 10) AND (c#42 >= 1)) AND (c#42 < 3)))], Format: parquet, Location: InMemoryFileIndex(1 paths)[path, PartitionFilters: [((c#42 = 0) OR ((c#42 >= 1) AND (c#42 < 3)))], PushedAggregation: [], PushedFilters: [Or(LessThan(a,10),GreaterThanOrEqual(a,10))], PushedGroupBy: [], ReadSchema: struct<a:bigint,b:bigint>, PushedFilters: [Or(LessThan(a,10),GreaterThanOrEqual(a,10))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: []
```

after
DataFilters pushed to ParquetScan is`[Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),LessThan(c,3)))]`
Filters pushed to ParquetInputFormat in partition `c=0` is `[LessThan(a,10)]`
Filters pushed to ParquetInputFormat in partition `c=1` or `c=2` is `[GreaterThanOrEqual(a,10)]`
```
== Physical Plan ==
*(1) Filter (((a#40L < 10) AND (c#42 = 0)) OR (((a#40L >= 10) AND (c#42 >= 1)) AND (c#42 < 3)))
+- *(1) ColumnarToRow
   +- BatchScan[a#40L, b#41L, c#42, d#43] ParquetScan DataFilters: [(((a#40L < 10) AND (c#42 = 0)) OR (((a#40L >= 10) AND (c#42 >= 1)) AND (c#42 < 3)))], Format: parquet, Location: InMemoryFileIndex(1 paths)[path, PartitionFilters: [((c#42 = 0) OR ((c#42 >= 1) AND (c#42 < 3)))], PushedAggregation: [], PushedFilters: [Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),Le..., PushedGroupBy: [], ReadSchema: struct<a:bigint,b:bigint>, PushedDynamicFilters: [Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),LessThan(c,3)))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: [] 
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
origin tests and add new test.

### CODE REVIEW

**Comment by stczwd** [comment]
cc @huaxingao @HyukjinKwon @dongjoon-hyun @cloud-fan

**Comment by huaxingao** [comment]
@stczwd Thanks for working on this. I haven't looked at your implementation yet. Based on the PR description, this PR will push down both data filters and partition filters, but it seems to me that we only need to push down data filters, because the partition filters have already been handled by partition pruning.

**Comment by huaxingao** [comment]
cc @cloud-fan

**Comment by stczwd** [comment]
> @stczwd Thanks for working on this. I haven't looked at your implementation yet. Based on the PR description, this PR will push down both data filters and partition filters, but it seems to me that we only need to push down data filters, because the partition filters have already been handled by partition pruning.

Thanks for your attension. Yeap, partition filters have already been handled by partition pruning. However, the data filters, which pushed down, contains conditions for every partitions, which means we need scan all conditions for each partition.
Simple examples.
1. if condition is `(a > 0 and c = 0) or c=2`, then no datafilter will be pushed down, we will scan all data of partition(c=0) and partition(c=1);
2. if condition is `(a > 10 and c=0) or (a <=10 and c=2)`, then the pushed filter will be `a>10 or a<=10`, we will also scan all data of partition(c=0) and partition(c=1);

**Comment by stczwd** [comment]
@huaxingao I have changed the PR description，and listed the difference between the two plans, so that the difference can be better seen

**Comment by stczwd** [comment]
Also, cc @sunchao and @MaxGekk.

**Comment by cloud-fan** [comment]
This looks like an interesting optimization. For file sources, the pushed predicate can be dynamic: we can fill the value of the partition columns in the pushed predicate, when pushing down the predicate to Parquet at the task runtime.

cc @sigmod @maryannxue @viirya @sunchao @c21

**Comment by viirya** [comment]
Hmm, the first glance looks a bit confused, actually. Maybe it is good to improve the PR description. So, I guess this means, if the data filter is partition related, e.g. a < 10 is only for partition c = 0, and a >= 10 is for partitions c >= 1 and c < 3, then once we strip out partition filter during pushing down, we will end with a < 0 or a >= 10 which actually scans all data.

But we only want a < 10 in partition c = 0, and a >= 10 in other two partitions.

Is it correct?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -557,6 +562,7 @@ class ParquetFilters(
         } else {
           Some(sources.Or(leftResultOptional.get, rightResultOptional.get))
         }
+
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:565]: unnecessary change.

**Comment by sunchao** [comment]
+1 on what @viirya said - the PR description is a bit confusing, for instance I don't really know what column `a` and `c` are for when first looking at it. It'd be better if you can also put the original SQL query there.

Also I'm curious how Parquet handles pushed filters like `(a < 10 and c == 0) or ((a >= 10 and c >= 1) and c < 3))`, since the partition column are not materialized.

**Comment by c21** [comment]
+1 to @sunchao and @viirya's points. Could we add the evidence of Parquet skipping files/row-groups (either a micro benchmark or some logs during execution or code pointers), when we push down partition filter here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala (hunk 1 above)
Comment (stczwd) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:565]: Just break a new line. Only for better code reading.

**Comment by stczwd** [comment]
> Hmm, the first glance looks a bit confused, actually. Maybe it is good to improve the PR description. So, I guess this means, if the data filter is partition related, e.g. a < 10 is only for partition c = 0, and a >= 10 is for partitions c >= 1 and c < 3, then once we strip out partition filter during pushing down, we will end with a < 0 or a >= 10 which actually scans all data.
> 
> But we only want a < 10 in partition c = 0, and a >= 10 in other two partitions.
> 
> Is it correct?

Yes, this is what I trying to do

**Comment by stczwd** [comment]
> +1 on what @viirya said - the PR description is a bit confusing, for instance I don't really know what column `a` and `c` are for when first looking at them. It'd be better if you can also put the original SQL query there.
> 
> Also I'm curious how Parquet handles pushed filters like `(a < 10 and c == 0) or ((a >= 10 and c >= 1) and c < 3))`, since the partition columns are not materialized.

Thanks for your attension, @sunchao @viirya @c21. Just as @cloud-fan said, we can push all filters, data filters and partition filters, to ParquetScan, and evaluate and prune partition filter with partition values at runtime, before pushed into parquet. With this, we can read `a < 10` in partition `c=0` and read `a>=10` in partition `c>=1 and c<2`, instead of reading `a<10 or a>=10` in partition `c=0` and partition `c>=1 and c<2`.

**Comment by stczwd** [comment]
@viirya @sunchao @c21 I have updated the PR description, does it look clear now?

**Comment by sunchao** [comment]
> .. and evaluate and prune partition filter with partition values at runtime, before pushed into parquet

Ah I see it now, thanks @stczwd .

**Comment by LuciferYang** [comment]
> > Could we add the evidence of Parquet skipping files/row-groups (either a micro benchmark or some logs during execution or code pointers), when we push down partition filter here?
> 
> @c21 I have add some benchmark tests in FilterPushdownBenchmark, and run them in github actions. Test code can be found [here](https://github.com/stczwd/spark/blob/SPARK-38041-2/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala#L81).
> 
> Test result
> 
> ```
> OpenJDK 64-Bit Server VM 1.8.0_322-b06 on Linux 5.11.0-1028-azure
> Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
> Data filter with partitions: ((a = 10 and part = 0) or (a = 10240 and part = 1) or (part = 2)):  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
> ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> Parquet Vectorized with partition                                                                        3039           3157         122          5.2         193.2       1.0X
> Parquet Vectorized with partition (Pushdown)                                                             1548           1568          15         10.2          98.4       2.0X
> 
> OpenJDK 64-Bit Server VM 1.8.0_322-b06 on Linux 5.11.0-1028-azure
> Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
> Data filter with partitions: ((a > 10 and part = 0) or (a <= 10 and part >=1 and part < 3)):  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
> ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> Parquet Vectorized with partition                                                                     2942           2997          40          5.3         187.1       1.0X
> Parquet Vectorized with partition (Pushdown)                                                          1497           1513          15         10.5          95.2       2.0X
> ```

@stczwd Can you add the benchmark code to this pr and use GA to produce the benchmark results?

**Comment by LuciferYang** [comment]
Also cc @dongjoon-hyun I'm not sure whether this optimization is also applicable to Orc file, and whether Orc also needs this optimization?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala (hunk 1 above)
Comment (stczwd) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: removed now

**Comment by c21** [comment]
I think the optimization should be able to add for ORC as well.

**Comment by stczwd** [comment]
> I think the optimization should be able to add for ORC as well.

Great. I will create a new JIRA for ORC optimization, this PR will only support parquet optimization.

**Comment by huaxingao** [comment]
@stczwd The PR makes sense to me. I have one question: does your optimization work for V1 file source? Seems to me that data filters and partition filters are separated differently in V1 and V2 file sources and your optimization only work for V2 file source?

**Comment by stczwd** [comment]
>  Seems to me that data filters and partition filters are separated differently in V1 and V2 file sources and your optimization only work for V2 file source?

Yes, this is only worked with v2 file source. V1 file source won't work even if `spark.sql.parquet.filterPushdown.partition` is set to `true`.

**Comment by stczwd** [comment]
> @stczwd Can you add the benchmark code to this pr and use GA to produce the benchmark results?

@LuciferYang The benchmark code and result generated by GA have been updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 2]
```diff
@@ -630,4 +631,35 @@ object PartitioningUtils extends SQLConfHelper{
...
+   * @return the result of evaluate result.
+   */
+  def evaluatePartitionFilter(
+      partitionSchema: StructType,
+      partitionValues: Option[InternalRow],
+      partitionFilter: sources.Filter): Boolean = {
+    val predicate =
+      StructFilters.filterToExpression(partitionFilter, StructFilters.toRef(partitionSchema))
+    if (predicate.isDefined) {
+      val boundPredicate = Predicate.createInterpreted(predicate.get.transform {
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:651]: Maybe we can extract line651 ~ 655 into a new function, then return 
```scala
...
predicate.isDefined && partitionValues.exists(newFunc.eval)
``` 
?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 3]
```diff
@@ -630,4 +631,35 @@ object PartitioningUtils extends SQLConfHelper{
...
+   * @param partitionFilter partition filter.
+   * @return the result of evaluate result.
+   */
+  def evaluatePartitionFilter(
+      partitionSchema: StructType,
+      partitionValues: Option[InternalRow],
+      partitionFilter: sources.Filter): Boolean = {
+    val predicate =
+      StructFilters.filterToExpression(partitionFilter, StructFilters.toRef(partitionSchema))
+    if (predicate.isDefined) {
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:650]: Maybe we can extract line651 ~ 655 into a new function, then return
```scala
...
predicate.isEmpty || predicate.isDefined && partitionValues.exists(newFunc.eval)
```

Is that right?

**Comment by LuciferYang** [comment]
From the results of the new benchmark, it can show some performance improvements, also ping @wangyum

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 4]
```diff
@@ -2053,6 +2053,34 @@ class ParquetV2FilterSuite extends ParquetFilterSuite {
       }
     }
   }
+
+  test("SPARK-38041: Data filter push down with partition filter") {
+    val data = spark.createDataFrame((1 to 4).map(i => Tuple1(Option(i))))
+    withNestedDataFrame(data).foreach { case (inputDF, colName, resultFun) =>
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:2059]: I think the goal of this test to to check if the partition filters are also pushed down. It doesn't matter if the data filter is on nested column or not, so no need to test this on different levels of nested columns.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 5]
```diff
@@ -2053,6 +2053,34 @@ class ParquetV2FilterSuite extends ParquetFilterSuite {
...
+      withTempPath { file =>
+
+        val partName = "part"
+        inputDF.withColumn(partName, lit(0))
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:2063]: Is this a partition column? Seems like a data column to me.

Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:2063]: Ah, I miss the `.partitionBy(partName)`, I will add it in next commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 4 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:2059]: Agreed. I will remove this in next commit.

**Comment by huaxingao** [comment]
@stczwd 
Thanks for the quick fix.
The optimization in your PR makes sense to me, but I think there is a problem. Use this filter `(id > 0 AND p = 0) OR (id = 2 AND p = 1)` as an example, the current physical plan is 
```
*(1) Filter (((id#9L > 0) AND (p#10 = 0)) OR ((id#9L = 2) AND (p#10 = 1)))
+- *(1) ColumnarToRow
   +- BatchScan[id#9L, p#10] ParquetScan DataFilters: [(((id#9L > 0) AND (p#10 = 0)) OR ((id#9L = 2) AND (p#10 = 1)))], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/private/var/folders/pt/_5f4sxy56x70dv9zpz032f0m0000gn/T/spark-51..., PartitionFilters: [((p#10 = 0) OR (p#10 = 1))], PushedAggregation: [], PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedGroupBy: [], ReadSchema: struct<id:bigint>, PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: []
```
This actually looks like a bug to me. I have fixed it to
```
== Physical Plan ==
*(1) Filter ((id#9L > 0) OR (id#9L = 2))
+- *(1) ColumnarToRow
   +- BatchScan[id#9L, p#10] ParquetScan DataFilters: [((id#9L > 0) OR (id#9L = 2))], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/private/var/folders/pt/_5f4sxy56x70dv9zpz032f0m0000gn/T/spark-38..., PartitionFilters: [((p#10 = 0) OR (p#10 = 1))], PushedAggregation: [], PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedGroupBy: [], ReadSchema: struct<id:bigint>, PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: []
```
I will submit a PR for the fix.

I think your PR needs the data filters to be ` [(((id#9L > 0) AND (p#10 = 0)) OR ((id#9L = 2) AND (p#10 = 1)))]`. 

After my fix, you can still do your optimization, but need to change some code .

**Comment by huaxingao** [comment]
cc @cloud-fan for the above comment

**Comment by LuciferYang** [comment]
Should `(id > 0 AND p = 0) OR (id = 2 AND p = 1)`  really be optimized to

```
== Physical Plan ==
*(1) Filter ((id#9L > 0) OR (id#9L = 2))
+- *(1) ColumnarToRow
   +- BatchScan[id#9L, p#10] ParquetScan DataFilters: [((id#9L > 0) OR (id#9L = 2))], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/private/var/folders/pt/_5f4sxy56x70dv9zpz032f0m0000gn/T/spark-38..., PartitionFilters: [((p#10 = 0) OR (p#10 = 1))], PushedAggregation: [], PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedGroupBy: [], ReadSchema: struct<id:bigint>, PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: []
```

It looks wrong for record `id=1, p=1`, or did I get it wrong @huaxingao?

**Comment by huaxingao** [comment]
@LuciferYang 
You are right. I got the post scan filter wrong. It should be
```
== Physical Plan ==
*(1) Filter (((id#9L > 0) AND (p#10 = 0)) OR ((id#9L = 2) AND (p#10 = 1)))
+- *(1) ColumnarToRow
   +- BatchScan[id#9L, p#10] ParquetScan DataFilters: [((id#9L > 0) OR (id#9L = 2))], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/private/var/folders/pt/_5f4sxy56x70dv9zpz032f0m0000gn/T/spark-38..., PartitionFilters: [((p#10 = 0) OR (p#10 = 1))], PushedAggregation: [], PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedGroupBy: [], ReadSchema: struct<id:bigint>, PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: []
```

**Comment by stczwd** [comment]
> After my fix, you can still do your optimization, but need to change some code .

Hm, I think you mean the [FileScanBuilder.pushFilters](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala#L71) should split the partitionFilter and dataFilter? And for parquet or orc, they should overwrite the method and get the filters before split? is it what you mean? @huaxingao

**Comment by huaxingao** [comment]
Yes, I mean [FileScanBuilder.pushFilters](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala#L71) should split the partitionFilter and dataFilter, and only data filter should be pushed down to parquet/orc.  But not splitting doesn't cause any issue anyways because later on Spark checks the filter and only creates parquet/orc filters for data filters. I will leave this as is so  you can use it for your optimization.

**Comment by huaxingao** [comment]
@stczwd 
What is the reason for your optimization not working with V1? is it because only data filters are pushed to the `FileScan`?
```
== Physical Plan ==
*(1) Filter (((id#9L > 0) AND (p#10 = 0)) OR ((id#9L = 2) AND (p#10 = 1)))
+- *(1) ColumnarToRow
   +- FileScan parquet [id#9L,p#10] Batched: true, DataFilters: [((id#9L > 0) OR (id#9L = 2))], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/private/var/folders/pt/_5f4sxy56x70dv9zpz032f0m0000gn/T/spark-7d..., PartitionFilters: [((p#10 = 0) OR (p#10 = 1))], PushedFilters: [Or(GreaterThan(id,0),EqualTo(id,2))], ReadSchema: struct<id:bigint>
```

**Comment by stczwd** [comment]
> @stczwd What is the reason for your optimization not working with V1? is it because only data filters are pushed to the `FileScan`?

Yes, the data filters that pushed to the FileScan and the way filter pushed down are different in V1 and V2. Thus I thought start a work on V2 first and then optimize V1 based on the same `ParquetFilters` later.

**Comment by cloud-fan** [comment]
How hard it is to support file source v1? This is a big PR and it's a pity if it only works for file source v2 which is off by default,

**Comment by stczwd** [comment]
> How hard it is to support file source v1? This is a big PR and it's a pity if it only works for file source v2 which is off by default,

@cloud-fan Thanks for your attension. Actually, there are not many changes in this PR, most of the changes are in the benchmark results.
As for supporting os file source v1,  we split dataFilters and partitionFilters at [FileSourceStrategy.apply](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L188), and we need a way to push whole filters to `Fileformat.buildReaderWithPartitionValues`. Not a big PR, but it can be used in ORC v1 file source. Maybe another PR is better?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 6]
```diff
@@ -617,6 +623,10 @@ class ParquetFilters(
     nameToParquetField.contains(name) && valueCanMakeFilterOn(name, value)
   }
 
+  private def canMakeFilterOnPart(filter: sources.Filter): Boolean = {
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:626]: Seems to me this method name is a bit misleading. We are not really making a Parquet Filter for this partition filter and push it down to Parquet. It might be better to call it something like `canDynamicallyPushDownDataFilter`.

**Comment by huaxingao** [comment]
```
== Physical Plan ==
*(1) Filter (((a#40L < 10) AND (c#42 = 0)) OR (((a#40L >= 10) AND (c#42 >= 1)) AND (c#42 < 3)))
+- *(1) ColumnarToRow
   +- BatchScan[a#40L, b#41L, c#42, d#43] ParquetScan DataFilters: [(((a#40L < 10) AND (c#42 = 0)) OR (((a#40L >= 10) AND (c#42 >= 1)) AND (c#42 < 3)))], Format: parquet, Location: InMemoryFileIndex(1 paths)[path, PartitionFilters: [((c#42 = 0) OR ((c#42 >= 1) AND (c#42 < 3)))], PushedAggregation: [], PushedFilters: [Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),Le..., PushedGroupBy: [], ReadSchema: struct<a:bigint,b:bigint>, PushedFilters: [Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),LessThan(c,3)))], PushedAggregation: [], PushedGroupBy: [] RuntimeFilters: [] 
```
Seems to me this plan might be a bit misleading. The data filters are actually `[((id#9L > 0) OR (id#9L = 2))]` and the predicate pushed down to Parquet is `[((id#9L > 0) OR (id#9L = 2))]`. With the pushed filter to be `[Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),LessThan(c,3)))]`, it makes me feel that we are constructing a Parquet predicate `[Or(And(LessThan(a,10),EqualTo(c,0)),And(And(GreaterThanOrEqual(a,10),GreaterThanOrEqual(c,1)),LessThan(c,3)))]` and push it down to Parquet.

**Comment by jackylee-ch** [comment]
> The data filters are actually [((id#9L > 0) OR (id#9L = 2))] and the predicate pushed down to Parquet is [((id#9L > 0) OR (id#9L = 2))]

Hm, acutally, the predicate pushed down to Parquet should be `[(id#9L > 0)]` or `[(id#9L = 2)]`, that's why we need runtime filter pruning.
BTW, it's the meaning of the pushed filter changed. In this PR, the pushed filter means the filter of push to runtime, not the filter of push to parquet. Maybe using `PushedRuntimeFilters` instande of `PushedFilters`?

**Comment by huaxingao** [comment]
> Maybe using `PushedRuntimeFilters` instande of `PushedFilters`

Sounds OK to me.
The property name is a bit confusing too. Maybe change it as well?

**Comment by jackylee-ch** [comment]
> Sounds OK to me. The property name is a bit confusing too. Maybe change it as well?

Not get what `property name` means. You mean this should be also changed in `ParquetScan.description` and `ParquetScan.getMetaData`?

**Comment by huaxingao** [comment]
Sorry I didn't make it clear. I feel `PARQUET_FILTER_PUSHDOWN_PARTITION_ENABLED` is a bit misleading. We are not really push down partition filters. What we do is to dynamically push down data filters based on the partition values. Shall we call this `PARQUET_FILTER_RUNTIME_PUSHDOWN_ENABLED` or `PARQUET_FILTER_DYNAMICALLY_PUSHDOWN_ENABLED`?

**Comment by jackylee-ch** [comment]
> Shall we call this `PARQUET_FILTER_RUNTIME_PUSHDOWN_ENABLED` or `PARQUET_FILTER_DYNAMICALLY_PUSHDOWN_ENABLED`?

Ah, I got it. The second one sounds good to me, I will change it too.

**Comment by jackylee-ch** [comment]
cc @huaxingao and @cloud-fan

**Comment by jackylee-ch** [comment]
@huaxingao any more question about this pr? can this pr be merged now?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -849,6 +849,17 @@ object SQLConf {
...
     .createWithDefault(true)
 
+  val PARQUET_FILTER_DYNAMICALLY_PUSHDOWN_ENABLED =
+    buildConf("spark.sql.parquet.filterPushdown.dynamically")
+      .doc("If true, enables Parquet data filter push-down optimization dynamically. " +
+        "Data filter will be pushed down to ParquetScan with partition Filters, " +
+        "and prune partition filters before pushing to parquet at runtime. " +
+        s"This configuration only has an effect when '${PARQUET_FILTER_PUSHDOWN_ENABLED.key}' " +
+        "is enabled.")
+      .version("3.3.0")
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:859]: Hmm... the master is already 3.4.0,  should change it here

**Review by LuciferYang — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 7 above)
Comment (jackylee-ch) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:859]: @huaxingao @cloud-fan @MaxGekk can this pr merged in 3.3.0? If not, I should change the jira.

**Comment by huaxingao** [comment]
@jackylee-ch Sorry I didn't have time to take a look yet. Will try to take a look during the weekend. It's probably a little rush to merge in 3.3.0.

**Comment by jackylee-ch** [comment]
> @jackylee-ch Sorry I didn't have time to take a look yet. Will try to take a look during the weekend. It's probably a little rush to merge in 3.3.0.

Okey, I'm just asking. It's ok to me that merging this pr to 3.4.0, I will change the affect versions.

**Comment by huaxingao** [comment]
@cloud-fan 
I think we probably need to reach a consensus here :)

before this patch, the physical plan has

```
Filter:
(((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))

ParquetScan DataFilters: 
[(((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))]

PartitionFilters: 
[((part#15 = 0) OR (part#15 = 1))]

PushedFilters: 
[Or(LessThan(id,2),GreaterThan(id,3))]
```

after this patch, the physical plan has

```
Filter:
(((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))

ParquetScan DataFilters: 
[(((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))]

PartitionFilters: 
[((part#15 = 0) OR (part#15 = 1))]

PushedDynamicalFilters: 
[Or(And(LessThan(id,2),EqualTo(part,0)),And(GreaterThan(id,3),EqualTo(part,1)))]
```
Please note that `PushedFilters` is changed to `PushedDynamicalFilters`. `[Or(And(LessThan(id,2),EqualTo(part,0)),And(GreaterThan(id,3),EqualTo(part,1)))]` means data filter id < 2 is dynamically pushed to partition 0 and data filter id > 3 is dynamically pushed to partition 1.

Does this look OK to you?

**Comment by LuciferYang** [comment]
At least I think this is OK. Only pushing down the filter belonging to the current partition is conducive to reducing the file scan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 7 above)
Comment (jackylee-ch) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I have changed the affect version from 3.3.0 to 3.4.0

**Comment by jackylee-ch** [comment]
> @cloud-fan I think we probably need to reach a consensus here :)
> 
> before this patch, the physical plan has
> 
> ```
> Filter:
> (((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))
> 
> ParquetScan DataFilters: 
> [(((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))]
> 
> PartitionFilters: 
> [((part#15 = 0) OR (part#15 = 1))]
> 
> PushedFilters: 
> [Or(LessThan(id,2),GreaterThan(id,3))]
> ```
> 
> after this patch, the physical plan has
> 
> ```
> Filter:
> (((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))
> 
> ParquetScan DataFilters: 
> [(((id#14 < 2) AND (part#15 = 0)) OR ((id#14 > 3) AND (part#15 = 1)))]
> 
> PartitionFilters: 
> [((part#15 = 0) OR (part#15 = 1))]
> 
> PushedDynamicalFilters: 
> [Or(And(LessThan(id,2),EqualTo(part,0)),And(GreaterThan(id,3),EqualTo(part,1)))]
> ```
> 
> Please note that `PushedFilters` is changed to `PushedDynamicalFilters`. `[Or(And(LessThan(id,2),EqualTo(part,0)),And(GreaterThan(id,3),EqualTo(part,1)))]` means data filter id < 2 is dynamically pushed to partition 0 and data filter id > 3 is dynamically pushed to partition 1.
> 
> Does this look OK to you?

Kindly remind @cloud-fan

**Comment by jackylee-ch** [comment]
cc @huaxingao @cloud-fan. This PR has been ready for a long time now, can we merge this PR?
Also cc @sunchao @viirya

**Comment by huaxingao** [comment]
@jackylee-ch Thanks for your patience. The approach looks good to me. I concern a bit about the logging, though.
```
PushedDynamicalFilters: 
[Or(And(LessThan(id,2),EqualTo(part,0)),And(GreaterThan(id,3),EqualTo(part,1)))]
```
It's not very clear, but I don't have an easy way to make it better.

Since this PR has quite some changes, I prefer to get some opinions from others.

**Comment by jackylee-ch** [comment]
@huaxingao Maybe remove the `Stale` tag?
Also cc @cloud-fan @sunchao
